controller/kongplugininstallation/image/image.go

```diff
@@ -0,0 +1,142 @@
+package image
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	ociv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/memory"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/credentials"
+
+	"github.com/kong/gateway-operator/modules/manager/metadata"
+)
+
+// FetchPluginContent fetches the content of the plugin from the image URL. When authentication is not needed pass nil.
+func FetchPluginContent(ctx context.Context, imageURL string, credentialsStore credentials.Store) ([]byte, error) {
+	ref, err := name.ParseReference(imageURL)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected format of image url: %w", err)
+	}
+	registryName, repositoryName, imageTag := ref.Context().RegistryStr(), ref.Context().RepositoryStr(), ref.Identifier()
+	// Errors for NewRegistry(..) and Repository(..) should never happen because the image URL has been already validated above.
+	registry, err := remote.NewRegistry(registryName)
+	if err != nil {
+		return nil, fmt.Errorf("for image: %s unexpected registry: %s, because: %w", imageURL, registryName, err)
+	}
+	var credentialFunc auth.CredentialFunc
+	if credentialsStore != nil {
+		credentialFunc = credentials.Credential(credentialsStore)
+	}
+	httpClient := auth.DefaultClient
+	httpClient.Header = map[string][]string{"User-Agent": {metadata.Metadata().UserAgent()}}
+	registry.Client = &auth.Client{
+		Client:     httpClient.Client,
+		Cache:      auth.NewCache(),
+		Credential: credentialFunc,
+	}
```

Comment:
`auth.DefaultClient` is a pointer to `auth.Client` so setting its `Header` field is ineffective in our case. We should explicitly set the `Header` field below when instantiating `auth.Client`.
```suggestion
	registry.Client = &auth.Client{
		Client:     auth.DefaultClient.Client,
		Header:     map[string][]string{"User-Agent": {metadata.Metadata().UserAgent()}},
		Cache:      auth.NewCache(),
		Credential: credentialFunc,
	}
```